src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -129,12 +132,52 @@ void FireEventsForDocument(DocumentId documentId)
             if (filename == null)
                 return;
 
-            if (!projectCodeModel.TryGetCachedFileCodeModel(filename, out var fileCodeModelHandle))
+            projectCodeModelAndFileNames.Add((projectCodeModel, filename));
+        }
+    }
+
+    private async Task FireEventsForChangedDocumentsAsync(
+        ArrayBuilder<(ProjectCodeModel projectCodeModel, string filename)> projectCodeModelAndFileNames,
+        CancellationToken cancellationToken)
+    {
+        // This logic preserves the previous behavior we had with IForegroundNotificationService.
+        // Specifically, we don't run on the UI thread for more than 15ms at a time. And we don't
+        // check the input queue more than once per ms. Once we have run for more than 15 ms,
+        // we wait 50ms before continuing.  These constants are just what we defined from
+        // legacy, and otherwise have no special meaning.
+        const int MaxTimeSlice = 15;
+        double nextInputCheckElapsedMs = 1;
+
+        var stopwatch = SharedStopwatch.StartNew();
+        foreach (var (projectCodeModel, filename) in projectCodeModelAndFileNames)
+        {
+            // If we've been asked to shutdown, don't bother reporting any more events.
+            if (_threadingContext.DisposalToken.IsCancellationRequested)
                 return;
 
-            var codeModel = fileCodeModelHandle.Object;
-            codeModel.FireEvents();
-            return;
+            FireEventsForDocument(projectCodeModel, filename);
+
+            // Keep firing events for this doc, as long as we haven't exceeded MaxTimeSlice ms or input isn't pending.
+            // We'll validate against those constraints at most once every 1 ms, to avoid spam checking the
+            // input queue, as this has shown up in performance profiles.
+            var elapsedMs = stopwatch.Elapsed.TotalMilliseconds;
+            if (elapsedMs > nextInputCheckElapsedMs)
+            {
+                if (elapsedMs > MaxTimeSlice || IsInputPending())
+                {
+                    await this.Listener.Delay(DelayTimeSpan.NearImmediate, cancellationToken).ConfigureAwait(true);
+                    stopwatch = SharedStopwatch.StartNew();
+                    elapsedMs = 0;
+                }
+
+                nextInputCheckElapsedMs = elapsedMs + 1;
+            }
```

Comment:
i assume this logic was the same as before.

Reply:
This logic changed, as IsInputPending showed up in the profiles